[NET_HANDLER] readonly attribute added to Param protos

AVSurfer123 opened the pull request:
Adding flag to the `Param` protobuf whether the param is read only for the student.
Also changed how UDP data is tested by directly checking the `DevData*` struct rather than doing string matching.

Closes #152

benliao1 commented:
Looks good, the only thing I'd add is maybe a test to make sure that the info is sent to Dawn?

AVSurfer123 commented:
TODO: Need to add assertion functions to `test.c` that will check the provided `DevData*` struct for correctness. This is since I removed the printing from the `net_handler_client` and moved it into the `net_handler_cli` where it belongs. I currently don't have the time to do this though if someone else wants to finish it.

levincent06 commented on tests/test.c:
```diff
@@ -489,6 +489,42 @@ void check_inputs(uint64_t expected_buttons, float expected_joysticks[4], robot_
+    print_pass();
+}
+
+void check_udp_device_param(DevData* dev_data, int dev_idx, char* param_name, param_type_t param_type, param_val_t param_val, uint8_t readonly) {
```
Is this function done? What is it used for? Should this be added to the header file?

AVSurfer123 replied:
I've finished this function now

levincent06 commented on tests/test.h:
```diff
@@ -83,6 +83,20 @@ void add_unordered_string_output(char* output);
+/******************** UDP Device Data Checks ******************/
+
+/**
+ * Checks that the given device was returned by the UDP thread.
```
Wdym by returned by the UDP thread? All device data are sent over UDP right?

AVSurfer123 replied:
I meant sent by the UDP thread. I will fix it

levincent06 commented on tests/test.h:
```diff
@@ -83,6 +83,20 @@ void add_unordered_string_output(char* output);
+ * 
+ * Args:
+ *  dev_data: Protobuf struct that is outputted by the UDP thread from get_next_device_data()
+ *  index: index where device should be
```
How does a test know which index a particular device would be in the protobuf?

AVSurfer123 replied:
They probably just need to run it out on their own using the CLI. it's also the same order as device indices in SHM, but with any gaps removed.

levincent06 commented on tests/test.h:
```diff
@@ -83,6 +83,20 @@ void add_unordered_string_output(char* output);
+ *  uid: what uid the device should be
+ * 
+ */
+void check_udp_device_exists(DevData* dev_data, int index, uint8_t type, uint64_t uid);
```
Would it make more sense to name this `check_device_data_sent()` or similar?

AVSurfer123 replied:
Well its already sent, but we are checking if a specific device was sent in that packet.

levincent06 commented on tests/client/net_handler_client.c:
```diff
@@ -302,11 +283,9 @@ void start_net_handler() {
         null_fp = fopen("/dev/null", "w");
 
         // init the mutex that will control whether udp prints to screen
```
"init the lock on the global variable that holds the most recent device data"?

levincent06 commented on tests/client/net_handler_client.h:
```diff
@@ -82,11 +82,10 @@ void send_challenge_data(robot_desc_field_t client, char** data, int num_challen
- * to be printed to standard output (since device data packets are constantly coming in, they
- * are normally suppressed; this function unsuppresses one single packet).
+ * Calling this function will return the most recent device data packet coming into Dawn from Runtime.
+ * NOTE: you must free the pointer returned by this function using dev_data__unpacked
```
Do you mean `dev_data__free_unpacked()`?

levincent06 commented on tests/client/net_handler_client.c:
```diff
@@ -4,18 +4,20 @@
-FILE* udp_output_fp = NULL;   // holds current output location of incoming UDP messages
 FILE* null_fp = NULL;         // file pointer to /dev/null
 
+DevData* device_data = NULL;
```
Maybe add a comment that this holds the most recent device data received

DanielMolina24 commented on tests/cli/net_handler_cli.c:
```diff
@@ -315,6 +315,35 @@ void prompt_reroute_output() {
     update_tcp_output_fp(dest);
 }
 
+void print_next_dev_data() {
```
Can you explain why you decided to move this function to the net_handler_cli? Personally, I always viewed the cli as  an area that displays values or takes in inputs, rather than getting the data and parsing it. I feel the majority of this should be within the client

AVSurfer123 replied:
Talked about this during a meeting some time ago.

benliao1 commented on net_handler/udp_conn.c:
```diff
@@ -94,6 +94,7 @@ static void* send_device_data(void* args) {
                     }
                     param__init(param);
                     param->name = device_info->params[j].name;
+                    param->readonly = device_info->params[j].write == 0;
```
Can we add a TODO comment or something after this to make sure we know to use `stdbool.h` once we get around to using that header? And to make it more explicit, I usually prefer writing it as
```
device_info->params[j].write ? 0 : 1;
```

AVSurfer123 replied:
I made it check for `false` now. I also don't like using the ternary operator for this.

benliao1 replied:
I think it's more explicit; also if you don't specifically set it to 1, the true value could be any nonzero value, which can cause problems up the line when packing the protobuf / interpreting the boolean value.

AVSurfer123 replied:
The C language will only return 1 or 0 for an == operation, so you don't need to worry about that.

benliao1 commented on tests/client/net_handler_client.c:
```diff
@@ -622,6 +600,17 @@ void send_device_subs(dev_subs_t* subs, int num_devices) {
     usleep(400000);  // allow time for net handler and runtime to react and generate output before returning to client
 }
 
+DevData* get_next_dev_data() {
```
Can you add some comments on and in this function to explain what it does?

benliao1 reviewed (commented):
Just a couple of small comments!

levincent06 commented on tests/client/net_handler_client.c:
```diff
@@ -80,62 +82,38 @@ static int connect_tcp(robot_desc_field_t client) {
 
 /**
  * Function to receive data from net_handler on the UDP port
```
Add a comment that it updates the global variable with the incoming DeviceData (Rn from the header it's hard to tell what the function output is)

AVSurfer123 replied:
Fixed

levincent06 commented on tests/integration/tc_71_18.c:
```diff
@@ -22,24 +22,33 @@ char dev2_header[] = "Device No. 1:\ttype = SimpleTestDevice, uid = 17185, itype
 
 /**
  * Helper function to send dev1_params subscriptions
  * to device 1 and dev2_params subscriptions to device 2
```
Add a comment mentioning that it returns a DeviceData message that *should* reflect the new subscription, and that this message should be checked

AVSurfer123 replied:
Added
